Tasks/Common/codecoverage-tools/cobertura/cobertura.ant.ccenabler.ts

```diff
@@ -0,0 +1,195 @@
+
+        tl.debug("Input parameters: " + JSON.stringify(ccProps));
+
+        _this.reportDir = "CCReport43F6D5EF";
+        _this.reportbuildfile = "CCReportBuildA4D283EG.xml";
+        _this.buildFile = ccProps["buildfile"];
+        let classFilter = ccProps["classfilter"];
+        let srcDirs = ccProps["sourcedirectories"];
+        if (str(srcDirs).isEmpty()) {
+            srcDirs = ".";
+        }
+        _this.sourceDirs = srcDirs;
+        _this.classDirs = ccProps["classfilesdirectories"];
+
+        let filter = _this.extractFilters(classFilter);
+        _this.excludeFilter = _this.applyFilterPattern(filter.excludeFilter).join(",");
+        _this.includeFilter = _this.applyFilterPattern(filter.includeFilter).join(",");
+
+        tl.debug("Reading the build file: " + _this.buildFile);
+
+        return util.readXmlFileAsJson(_this.buildFile).
+            then(function (resp) {
+                return _this.addCodeCoverageData(resp);
+            })
+            .thenResolve(true);
+    }
+
+    protected applyFilterPattern(filter: string): string[] {
+        let ccfilter = [];
+        let _this = this;
+
+        if (!util.isNullOrWhitespace(filter)) {
+            str(util.trimToEmptyString(filter)).replaceAll(".", "/").s.split(":").forEach(exFilter => {
+                if (exFilter) {
+                    ccfilter.push(str(exFilter).endsWith("*") ? ("**/" + exFilter + "/**") : ("**/" + exFilter + ".class"));
+                }
+            });
+        }
+
+        tl.debug("Applying the filter pattern: " + filter + " op: " + ccfilter);
+        return ccfilter;
+    }
+
+    protected getClassData(): any {
+        let _this = this;
+        let fileset = [];
+        let classDirs = _this.classDirs;
+
+        if (str(classDirs).isEmpty()) {
+            classDirs = ".";
+        }
+        classDirs.split(",").forEach(cdir => {
+            let filter = {
+                $: {
+                    dir: cdir,
+                    includes: _this.includeFilter,
+                    excludes: _this.excludeFilter
+                }
+            };
+            fileset.push(filter);
+        });
+        return fileset;
+    }
+
+    protected createReportFile(reportContent: string): Q.Promise<void> {
+        let _this = this;
+        tl.debug("Creating the report file: " + _this.reportbuildfile);
+
+        let reportFile = path.join(path.dirname(_this.buildFile), _this.reportbuildfile);
+        return util.writeFile(reportFile, reportContent);
+    }
+
+    protected addCodeCoverageData(pomJson: any): Q.Promise<any[]> {
+        let _this = this;
+
+        if (!pomJson || !pomJson.project) {
+            return Q.reject<any>(tl.loc("InvalidBuildFile"));
+        }
+
+        let reportPluginData = ccc.coberturaAntReport(_this.sourceDirs, path.join(path.dirname(_this.buildFile), _this.reportDir));
+        return Q.all([_this.addCodeCoverageNodes(pomJson), _this.createReportFile(reportPluginData)]);
```

Comment:
Just be aware if there's async I/O down those paths Q.all can do parallel I believe.  Should this just be called sequentially?